File: src/lib/worship.ts
// Family Worship (turn 11): a guided order of service, morning and evening.
// The liturgy is data — an ordered list of typed elements (scripture, prayer,
// song, creed, prompts, and one catechism hand-off) — so the eight step
// screens are a single renderer and the AM/PM services are two JSON files,
// not two flows. Scripture here is historic public-domain text authored in
// the content files (the canvas's "Sourcing" note), never fetched.
// Statically imported so the bundler traces the data — a dynamic fs read
// breaks in Vercel's serverless bundles (see programContent.ts).
import morningJson from '../../content/worship/morning.json';
import eveningJson from '../../content/worship/evening.json';

export type Daypart = 'morning' | 'evening';

/** Where an element points beyond its own text — a recording of the song, the
 * book a prayer originates in. Informational detours only, never required. */
export interface ElementSource {
  label: string;
  href: string;
  /** true renders ↗ (leaves the app); false/absent renders › (stays in it). */
  external?: boolean;
}

export type WorshipElement = { lede?: string; source?: ElementSource } & (
  | { type: 'scripture'; citation: string; osis: string; text: string }
  | { type: 'prayer'; text: string; amen?: boolean }
  | { type: 'song'; title: string; lyrics: string }
  | { type: 'creed'; title: string; text: string }
  | { type: 'catechism' }
  | { type: 'instruction'; text: string }
  | { type: 'prompts'; items: { label: string; text: string }[] }
);

/** An element position whose content rotates by day — the daily psalm, the
 * reading, the song. Fixed sibling elements stay fixed. */
type RotationElement = { type: 'rotation'; options: WorshipElement[] };

interface LiturgyStepJson {
  role: string;
  elements: (WorshipElement | RotationElement)[];
}

export interface WorshipStep {
  role: string;
  elements: WorshipElement[];
}

export interface WorshipService {
  daypart: Daypart;
  steps: WorshipStep[];
}

const LITURGIES: Record<Daypart, LiturgyStepJson[]> = {
  morning: morningJson as unknown as LiturgyStepJson[],
  evening: eveningJson as unknown as LiturgyStepJson[],
};

/** Whole days since the epoch for a local calendar date — the deterministic
 * index rotating slots key off, so the whole household sees one service per
 * day regardless of the hour they gather. */
export const dayIndexOf = (date: Date): number => Math.floor(
  Date.UTC(date.getFullYear(), date.getMonth(), date.getDate()) / 86400000,
);

const resolveElement = (
  element: WorshipElement | RotationElement,
  dayIndex: number,
): WorshipElement | null => {
  if (element.type !== 'rotation') return element;
  if (element.options.length === 0) return null;
  return element.options[dayIndex % element.options.length];
};

/** The concrete service for one daypart on one date, rotations resolved. */
export const getService = (daypart: Daypart, date: Date): WorshipService => {
  const dayIndex = dayIndexOf(date);
  return {
    daypart,
    steps: LITURGIES[daypart].map((step) => ({
      role: step.role,
      elements: step.elements
        .map((element) => resolveElement(element, dayIndex))
        .filter((element): element is WorshipElement => element !== null),
    })),
  };
};

/** The order-preview detail line for a step (mockup 11a): what's named after
 * the em-dash, e.g. "Call to Worship — Psalm 100:2". Null when the role
 * stands alone; the catechism step's question detail is the caller's to fill
 * (it depends on the child's track, not the liturgy). */
export const stepDetail = (step: WorshipStep): string | null => {
  const lead = step.elements[0];
  if (!lead) return null;
  if (lead.type === 'scripture') return lead.citation;
  if (lead.type === 'song') return lead.title;
  if (lead.type === 'creed') return lead.title;
  return null;
};

File: ios/DomainKit/Sources/DomainKit/Worship.swift
// Ported from src/lib/worship.ts's element/step types (the daypart rotation
// machinery itself — dayIndexOf, getService — is Family Worship's own and
// out of scope here; a Devotion's steps are always pre-resolved, no
// rotation elements).
import Foundation

public struct ElementSource: Codable, Equatable, Sendable {
    public let label: String
    public let href: String
    /// true renders "leaves the app"; false/absent stays in it.
    public let external: Bool?
}

public struct PromptItem: Codable, Equatable, Sendable {
    public let label: String
    public let text: String
}

public enum WorshipElement: Equatable, Sendable {
    case scripture(lede: String?, citation: String, osis: String, text: String, source: ElementSource?)
    case prayer(lede: String?, text: String, amen: Bool, source: ElementSource?)
    case song(lede: String?, title: String, lyrics: String, source: ElementSource?)
    case creed(lede: String?, title: String, text: String, source: ElementSource?)
    case catechism(lede: String?)
    case instruction(text: String)
    case prompts(lede: String?, items: [PromptItem])

    public var typeName: String {
        switch self {
        case .scripture: return "scripture"
        case .prayer: return "prayer"
        case .song: return "song"
        case .creed: return "creed"
        case .catechism: return "catechism"
        case .instruction: return "instruction"
        case .prompts: return "prompts"
        }
    }
}

extension WorshipElement: Decodable {
    private enum CodingKeys: String, CodingKey {
        case type, lede, citation, osis, text, amen, title, lyrics, source, items
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let type = try container.decode(String.self, forKey: .type)
        let lede = try container.decodeIfPresent(String.self, forKey: .lede)
        let source = try container.decodeIfPresent(ElementSource.self, forKey: .source)
        switch type {
        case "scripture":
            self = .scripture(
                lede: lede,
                citation: try container.decode(String.self, forKey: .citation),
                osis: try container.decode(String.self, forKey: .osis),
                text: try container.decode(String.self, forKey: .text),
                source: source
            )
        case "prayer":
            self = .prayer(
                lede: lede,
                text: try container.decode(String.self, forKey: .text),
                amen: try container.decodeIfPresent(Bool.self, forKey: .amen) ?? false,
                source: source
            )
        case "song":
            self = .song(
                lede: lede,
                title: try container.decode(String.self, forKey: .title),
                lyrics: try container.decode(String.self, forKey: .lyrics),
                source: source
            )
        case "creed":
            self = .creed(
                lede: lede,
                title: try container.decode(String.self, forKey: .title),
                text: try container.decode(String.self, forKey: .text),
                source: source
            )
        case "catechism":
            self = .catechism(lede: lede)
        case "instruction":
            self = .instruction(text: try container.decode(String.self, forKey: .text))
        case "prompts":
            self = .prompts(lede: lede, items: try container.decode([PromptItem].self, forKey: .items))
        default:
            throw DecodingError.dataCorruptedError(forKey: .type, in: container, debugDescription: "Unknown element type \(type)")
        }
    }
}

public struct WorshipStep: Decodable, Equatable, Sendable {
    public let role: String
    public let elements: [WorshipElement]
}

/// The order-of-service preview line (15c/16a): a step's lead element's
/// citation/title, or nil when nothing worth previewing leads it.
public func stepDetail(_ step: WorshipStep) -> String? {
    guard let lead = step.elements.first else { return nil }
    switch lead {
    case .scripture(_, let citation, _, _, _): return citation
    case .song(_, let title, _, _): return title
    case .creed(_, let title, _, _): return title
    default: return nil
    }
}
